openedx/openedx-platform
# feat: add paginated HomePageCoursesV2 view with filtering & ordering

## Description

This PR adds a new version of the HomePageCourses API with pagination, filtering & ordering, intending to be used by the new Course Home MFE. The new API is based on HomePageCourses API with tweaks, so it gets courses with Course Overview helpers that allow filtering and ordering with query set API methods.

The scope of this implementation only considers the API endpoint used to list courses in the Course Home MFE which is the `HomePageCoursesViewV2` that serves `api/contentstore/v2/home/courses`. Other API endpoints in V1 will not be implemented as part of this initiative.

### API changes compared to V1

As I mentioned before, the HomePage Courses API version 2 is implemented based on the API version 1 with a few changes:

- We're using a new [get_course_context_v2](https://github.com/openedx/edx-platform/blob/908753b895f41d1f8c5dc3b6c34c26744dd50fa2/cms/djangoapps/contentstore/utils.py#L1526-L1566) function to get the list of active and archived courses, and the in process course actions. That function is based on [get_course_context](https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/utils.py#L1483-L1523) but with pagination configurations.

- Also, courses are not split into active/inactive but filtered using a status query parameter.

### Other changes

- When using the API, the `ENABLE_HOME_PAGE_COURSE_API_V2` feature toggle should be enabled, so we're not hitting Mongo to get all course summaries. Instead, we're listing them using the Django ORM for SQL db engines: https://github.com/openedx/edx-platform/blob/908753b895f41d1f8c5dc3b6c34c26744dd50fa2/cms/djangoapps/contentstore/views/course.py#L445. These changes make querying courses easier using Q objects instead of list filtering.
- Helpers modifications to support query params to ordering or filtering courses when needed: https://github.com/openedx/edx-platform/pull/34173/files#diff-937adee5129f53ead6b3c7c988d70db9e9e95862f17a6f85a1eea2388d12b461R418-R428

## Supporting information

For more information on this enhancement, please check:
https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4032856101/Studio+Home+incremental+upgrades+-+product+approach

## Testing instructions
1st, turn on the Django ORM only reads by setting:  `FEATURES["ENABLE_HOME_PAGE_COURSE_API_V2"] = True` in your environment. Then, login as an administrator or as course staff (they should both work):
1. You can use your browser to get your list of courses from Studio `HTTP(s)://<CMS_HOST>/api/contentstore/v2/home/courses`. Check that the output is the same as: `HTTP(s)://<CMS_HOST>/api/contentstore/v1/home/courses`
2. Now, try filtering courses by their course ID, org or display name: `/api/contentstore/v2/home/courses?search=demo`
4. Then, try ordering courses by their course ID, org or display name `/api/contentstore/v2/home/courses?order=display_name` , `/api/contentstore/v2/home/courses?order=-display_name` for alphabetic order, order=created or order=-created for creation date
Or both 3. and 4.

## Deadline

This effort is part of the Spanish consortium project, so it'd be ideal to merge this before the end of the project.
